File: Prim_algorithm.cc
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

// Structure to represent an edge in the graph
struct Edge {
    int src, dest, weight;
};

// Structure to represent a node in the adjacency list
struct Node {
    int vertex, weight;
};

class Graph {
private:
    vector<vector<Node>> adjacencyList; // Adjacency list to store the graph
    int numVertices;

public:
    Graph(int numVertices) {
        this->numVertices = numVertices;
        adjacencyList.resize(numVertices);
    }

    void addEdge(int src, int dest, int weight) {
        Node node;
        node.vertex = dest;
        node.weight = weight;
        adjacencyList[src].push_back(node);

        node.vertex = src; // For undirected graph
        adjacencyList[dest].push_back(node);
    }

    void primMST() {
        priority_queue<Node, vector<Node>, greater<Node>> pq; // Priority queue to store nodes
        vector<bool> visited(numVertices, false); // Keep track of visited nodes
        int src = 0; // Start with the first vertex as the source

        // Create a vector to store the MST
        vector<Edge> MST;

        // Mark the source node as visited and push it into the priority queue
        visited[src] = true;
        for (const auto& node : adjacencyList[src]) {
            pq.push(node);
        }

        while (!pq.empty()) {
            Node current = pq.top();
            pq.pop();

            int currentVertex = current.vertex;
            if (visited[currentVertex])
                continue;

            visited[currentVertex] = true;

            Edge edge;
            edge.src = currentVertex;
            edge.dest = current.weight;
            edge.weight = current.weight;
            MST.push_back(edge);

            // Add the adjacent nodes of the current node to the priority queue
            for (const auto& node : adjacencyList[currentVertex]) {
                pq.push(node);
            }
        }

        // Print the Minimum Spanning Tree
        cout << "Minimum Spanning Tree:" << endl;
        for (const auto& edge : MST) {
            cout << edge.src << " -- " << edge.dest << " with weight " << edge.weight << endl;
        }
    }
};

int main() {
    // Create a graph with 6 vertices
    Graph graph(6);

    // Add edges to the graph
    graph.addEdge(0, 1, 4);
    graph.addEdge(0, 2, 3);
    graph.addEdge(1, 2, 1);
    graph.addEdge(1, 3, 2);
    graph.addEdge(2, 3, 4);
    graph.addEdge(3, 4, 2);
    graph.addEdge(4, 5, 6);

    // Find and print the Minimum Spanning Tree.
    graph.primMST();

    return 0;
}

File: Prim algorithm.py
class Graph:
    def __init__(self):
        self.graph = {}

    # Undirected weighted graph using an adjacency list.
    def add_edge(self, Vertex_u, Vertex_v, nodes_weight):
        if Vertex_u not in self.graph:
            self.graph[Vertex_u] = []
        if Vertex_v not in self.graph:
            self.graph[Vertex_v] = []

        self.graph[Vertex_u].append((Vertex_v, nodes_weight))
        self.graph[Vertex_v].append((Vertex_u, nodes_weight))


# The Prim algorithm for computing the Minimum Spanning Tree of a graph

import heapq
def prim(graph):
    # Picking the starting vertex
    start_vertex = list(graph.keys())[0]

    # Begin by setting up the minimum spanning tree and marking the visited vertices.
    minimum_spanning_tree = []
    visited = {start_vertex}

    # Create a heap to store the edges and their weights
    edges = [(nodes_weight, start_vertex, neighbor) for neighbor, nodes_weight in graph[start_vertex]]
    heapq.heapify(edges)

    while edges:
        # Pop the edge with the minimum nodes_weight
        nodes_weight, Vertex_u, Vertex_v = heapq.heappop(edges)

        if Vertex_v not in visited:
            # Add the edge to the minimum spanning tree
            minimum_spanning_tree.append((Vertex_u, Vertex_v, nodes_weight))
            visited.add(Vertex_v)

            # Add the adjacent edges of Vertex_v to the heap
            for neighbor, nodes_weight in graph[Vertex_v]:
                if neighbor not in visited:
                    heapq.heappush(edges, (nodes_weight, Vertex_v, neighbor))

    return minimum_spanning_tree


# Create a graph
graph = Graph()
graph.add_edge('A', 'B', 4)
graph.add_edge('A', 'C', 8)
graph.add_edge('B', 'C', 2)
graph.add_edge('B', 'D', 5)
graph.add_edge('C', 'D', 1)
graph.add_edge('C', 'E', 6)
graph.add_edge('D', 'E', 3)
graph.add_edge('D', 'F', 4)
graph.add_edge('E', 'F', 2)

# Compute the MST using Prim's algorithm
minimum_spanning_tree = prim(graph.graph)

# Print the minimum spanning tree
for edge in minimum_spanning_tree:
    u, v, weight = edge
    print(f"{u} -- {v}: {weight}")